### PR TITLE
Extract Base Image for V4

### PR DESCRIPTION
**Pull request by CooperLink**
### PR information
Continuation of https://github.com/Azure/azure-functions-docker/pull/516 which is an effort to have all language images rely on a base host image. This effort will simplify updates to the docker images used in production and should reduce the need for large and difficult to parse PRs in this repo. 

This PR specifically enables dotnet images and brings the changes verified in v3 over to v4. I also updated the release pipeline to upload a zip file of the dockerfiles generated during a release so that we can trace the contents of each image in prod. 

- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/development-docs/cleaning-up-commits.md).
- [x] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and CI is passing.

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -27,9 +27,8 @@ jobs:
         git fetch --all
         git checkout -b ${{ github.event.inputs.targetBranch }} origin/${{ github.event.inputs.targetBranch }}
         git merge dev
-        ./host/3.0/buster/amd64/base/generate-composite.sh -r all
-        git add . 
-        git commit -m "Generated Images for Release"
+        ./host/generate-composite.sh -3 -r all
+        zip -r version3images.zip ./host/3.0/buster/amd64/out/
```
Comment (CooperLink) [inline .github/workflows/release.yml]: Just noticed that this is wrong. need to target the /release/ dir

File: host/4/bullseye/amd64/base/host.Dockerfile [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
 RUN BUILD_NUMBER=$(echo ${HOST_VERSION} | cut -d'.' -f 4) && \
     git clone --branch v${HOST_VERSION} https://github.com/Azure/azure-functions-host /src/azure-functions-host && \
     cd /src/azure-functions-host && \
     HOST_COMMIT=$(git rev-list -1 HEAD) && \
     dotnet publish -v q /p:BuildNumber=$BUILD_NUMBER /p:CommitHash=$HOST_COMMIT src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj -c Release --output /azure-functions-host --runtime linux-x64 && \
     mv /azure-functions-host/workers /workers && mkdir /azure-functions-host/workers && \
-    rm -rf /root/.local /root/.nuget /src
+    rm -rf /root/.local /root/.nuget /src \;
 
 RUN EXTENSION_BUNDLE_VERSION=1.8.1 && \
```
Comment (pragnagopa) [inline host/4/bullseye/amd64/base/host.Dockerfile]: Please remove Extension Bundles V1

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -27,9 +27,8 @@ jobs:
         git fetch --all
         git checkout -b ${{ github.event.inputs.targetBranch }} origin/${{ github.event.inputs.targetBranch }}
         git merge dev
-        ./host/3.0/buster/amd64/base/generate-composite.sh -r all
-        git add . 
-        git commit -m "Generated Images for Release"
+        ./host/generate-composite.sh -3 -r all
+        zip -r version3images.zip ./host/3.0/buster/amd64/release/
         git tag ${{ github.event.inputs.version }}
         git push
```
Comment (pragnagopa) [inline .github/workflows/release.yml:31]: rebase on dev? these changes should not part of this PR?

File: host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # Build the runtime from source
-ARG HOST_VERSION=3.4.1
+ARG HOST_VERSION=3.3.1
 FROM mcr.microsoft.com/azure-functions/base:grpc-2.27-arm32v7 as grpc-image
```
Comment (pragnagopa) [inline host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile]: rebase?

Comment (CooperLink) [reply host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile]: Will do :)

File: host/4/bullseye/amd64/base/host.Dockerfile [hunk 5]
```diff
@@ -15,16 +15,7 @@ RUN BUILD_NUMBER=$(echo ${HOST_VERSION} | cut -d'.' -f 4) && \
...
-    EXTENSION_BUNDLE_FILENAME=Microsoft.Azure.Functions.ExtensionBundle.1.8.1_linux-x64.zip && \
-    apt-get update && \
-    apt-get install -y gnupg wget unzip && \
-    wget https://functionscdn.azureedge.net/public/ExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION/$EXTENSION_BUNDLE_FILENAME && \
-    mkdir -p /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION && \
-    unzip /$EXTENSION_BUNDLE_FILENAME -d /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION && \
-    rm -f /$EXTENSION_BUNDLE_FILENAME && \
-    find /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION/bin/runtimes/ -mindepth 1 -type d -not -name "linux-x64" -prune -exec rm -rf {} + && \
-    EXTENSION_BUNDLE_VERSION_V2=2.8.3 && \
+RUN EXTENSION_BUNDLE_VERSION_V2=2.8.3 && \
```
Comment (CooperLink) [inline host/4/bullseye/amd64/base/host.Dockerfile]: All v4 builds should have run as a result of this commit. Bug.

File: host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile (hunk 4 above)
Comment (CooperLink) [reply host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile]: Rebased!

**Comment by pragnagopa** [comment]
V4 pipeline builds failed - @CooperLink - please take a look

File: host/4/bullseye/amd64/base/host.Dockerfile (hunk 5 above)
Comment (CooperLink) [reply host/4/bullseye/amd64/base/host.Dockerfile]: Fairly certain this has been fixed. Keeping an eye out on it though

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -41,11 +39,22 @@ jobs:
...
+      run: |
+        ./host/generate-composite.sh -3 -r all
+        zip -r ${{ github.event.inputs.version }}.zip ./host/3.0/buster/amd64/release/
+    - name : Generate v4 Assets
+      if : startsWith(github.event.inputs.version, '4')
+      run: |
+        ./host/generate-composite.sh -4 -r all
+        zip -r ${{ github.event.inputs.version }}.zip ./host/4/bullseye/amd64/release/
+    - name: Attach Assets to Release
       uses: actions/upload-release-asset@v1
```
Comment (TsuyoshiUshio) [inline .github/workflows/release.yml:53]: I'd happy to see you replace the code to a good GitHub action. :)

File: host/3.0/dotnet-build.yml [hunk 7]
```diff
@@ -35,6 +35,12 @@ steps:
     continueOnError: false
     env:
       pswd: $(dockerPassword)
+    
+  - bash: |
+      ./host/generate-composite.sh -3 dotnet
```
Comment (TsuyoshiUshio) [inline host/3.0/dotnet-build.yml:40]: `-3` is not intuitive.  How about `-v 3` or `-v 4`  The following code sample might help. 
https://github.com/Azure-Samples/openhack-devops-proctor/blob/main/provision-team/deploy.sh#L36-L50

File: host/4/bullseye/amd64/base/sharedconfig/start.sh [hunk 8]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
+
+export DOTNET_USE_POLLING_FILE_WATCHER=true
```
Comment (TsuyoshiUshio) [inline host/4/bullseye/amd64/base/sharedconfig/start.sh:3]: I'm curious what does it mean. Not review, just my question.

File: host/4/bullseye/amd64/base/sharedconfig/start.sh [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  export ASPNETCORE_URLS=http://*:$PORT
+fi
+
+if [ -z $SSH_PORT ]; then
+  export SSH_PORT=2222
+fi
+
+if [ "$APPSVC_REMOTE_DEBUGGING" == "TRUE" ]; then
+    export languageWorkers__node__arguments="--inspect=0.0.0.0:$APPSVC_TUNNEL_PORT"
+    export languageWorkers__python__arguments="-m ptvsd --host localhost --port $APPSVC_TUNNEL_PORT"
```
Comment (TsuyoshiUshio) [inline host/4/bullseye/amd64/base/sharedconfig/start.sh:17]: Java and Powershell doesn't support it? I don't know the powershell one, however, for java, https://github.com/Azure/azure-functions-java-worker/blob/dev/CONTRIBUTING.md#setting-up-the-end-to-end-development-environment

File: host/4/node-build.yml [hunk 10]
```diff
@@ -30,6 +32,12 @@ steps:
     continueOnError: false
     env:
       pswd: $(dockerPassword)
+  
+  - bash: |
+      ./host/generate-composite.sh -4 node
```
Comment (TsuyoshiUshio) [inline host/4/node-build.yml:37]: How about adding PublishArtifact task for pushing the generated Dockerfile for debugging?

**Review by TsuyoshiUshio — COMMENTED** [review]
I leave the comments, if you don't agree with me, I'd happy to approve. It is great work! Thank you Cooper!

File: host/4/bullseye/amd64/base/sharedconfig/start.sh (hunk 9 above)
Comment (CooperLink) [reply host/4/bullseye/amd64/base/sharedconfig/start.sh:17]: I just copied over the start.sh that is shared between all languages right now
I am not sure why we don't provide java args here for remote debugging.  Worth looking into after this PR

File: host/4/node-build.yml (hunk 10 above)
Comment (CooperLink) [reply host/4/node-build.yml:37]: This is a great idea. I am just not sure how we would publish that artifact or where it should go. The build pipelines should tell us if any images failed building.  But, it would be nice to have slightly higher traceability if there was some way to surface that artifact on the pipeline run of build that happens.

File: host/4/bullseye/amd64/base/sharedconfig/start.sh (hunk 8 above)
Comment (CooperLink) [reply host/4/bullseye/amd64/base/sharedconfig/start.sh:3]: This script? Or the #!/usr/bin/env bash?  
Script starts host. 
The other thing tells linux what style of bash to expect in the script

File: host/3.0/dotnet-build.yml (hunk 7 above)
Comment (CooperLink) [reply host/3.0/dotnet-build.yml:40]: I agree. And I actually spent several hours trying to get that flag to look like you mentioned. But, I had a hard time with order of input vars when I implemented it similarly to the link. Which I thought was harder to deal with than the number flags. 
If you could assist me in figuring it out I would be happy to change it. Just wasn't able to get shift working how I wanted locally.

File: host/4/bullseye/amd64/base/sharedconfig/start.sh (hunk 8 above)
Comment (TsuyoshiUshio) [reply host/4/bullseye/amd64/base/sharedconfig/start.sh:3]: No this option `DOTNET_USE_POLLING_FILE_WATCHER` mean. :)

File: host/4/bullseye/amd64/base/sharedconfig/start.sh (hunk 9 above)
Comment (TsuyoshiUshio) [reply host/4/bullseye/amd64/base/sharedconfig/start.sh:17]: Let me ask with @amamounelsayed and @AnatoliB . Do you know why we don't have it for java and powershell?

File: host/4/node-build.yml (hunk 10 above)
Comment (TsuyoshiUshio) [reply host/4/node-build.yml:37]: This might help. https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/utility/publish-build-artifacts?view=azure-devops

File: host/4/bullseye/amd64/base/sharedconfig/start.sh (hunk 9 above)
Comment (TsuyoshiUshio) [reply host/4/bullseye/amd64/base/sharedconfig/start.sh:17]: However, you don't need to wait the answer. as you said, we can consider following PR.

Comment (pragnagopa) [reply host/4/bullseye/amd64/base/sharedconfig/start.sh:17]: These settings are needed to enable remote debugging for Node / Python - AFAIK we do not have support for remote debugging Java applications

cc @ahmelsayed - as he has context on these settings as well.

File: host/4/node-build.yml (hunk 10 above)
Comment (CooperLink) [reply host/4/node-build.yml:37]: This is great Tsuyoshi. I think that I can and should use this to produce artifacts during build pipelines for each language. I have captured your comment in an issue on the repo here. 
https://github.com/Azure/azure-functions-docker/issues/560

Feels like something that should be addressed in a follow up pr but I agree that it is useful

File: host/3.0/dotnet-build.yml (hunk 7 above)
Comment (CooperLink) [reply host/3.0/dotnet-build.yml:40]: Captured this in my notes for the future. Next time we have our call we can work through it together?

**Review by TsuyoshiUshio — APPROVED** [review]
LGTM

File: host/generate-composite.sh [hunk 11]
```diff
@@ -5,6 +5,21 @@
...
+    ## Selects the directory of the version to be generated
+    ## Defaults to v3 for now 
+    if [ "$VERSION" == "4" ]; then
+        echo "Version 4 selected"
+        DIR=$(dirname $0)/4/bullseye/amd64/base
+    elif [ "$VERSION" == "3" ]; then
+        echo "Version 4 selected"
+        DIR=$(dirname $0)/3.0/buster/amd64/base
+    else
+        echo "No valid version selected. Defaulting to v3"
```
Comment (balag0) [inline host/generate-composite.sh]: thinking it might be better to not use defaults and throw if the argument is not set. avoids having to update the default at a later time and things potentially breaking silently.
since we control the caller side, we can make sure they  always pass explicit version.

Comment (CooperLink) [reply host/generate-composite.sh]: Makes sense. I will remove the default version so that it throws. And run a quick test

**Review by balag0 — APPROVED** [review]

**Review by balag0 — APPROVED** [review]

File: host/generate-composite.sh (hunk 11 above)
Comment (CooperLink) [reply host/generate-composite.sh]: Resolving this comment as I was able to implement and test. Thanks Bala

File: host/4/bullseye/amd64/base/sharedconfig/start.sh (hunk 8 above)
Comment (balag0) [reply host/4/bullseye/amd64/base/sharedconfig/start.sh:3]: if set dotnet filewatcher will actively poll for filesystem changes, if not set, it will rely on file change notifications (which doesnt always work in some disk setups)

File: host/4/bullseye/amd64/base/sharedconfig/start.sh (hunk 9 above)
Comment (AnatoliB) [reply host/4/bullseye/amd64/base/sharedconfig/start.sh:17]: Does not apply to PowerShell: even though PowerShell supports remote debugging, it is based on named pipes, not ports. We may be able to make it work end-to-end eventually, but this is certainly not in this PR scope.
